controller/DeliveryTool.php

```diff
@@ -193,14 +203,33 @@ protected function getLearnerUrl(\core_kernel_classes_Resource $delivery, Delive
 
         /** @var LtiAssignment $assignmentService */
         $assignmentService = $this->getServiceLocator()->get(LtiAssignment::SERVICE_ID);
-        if ($assignmentService->isDeliveryExecutionAllowed($delivery->getUri(), $user)) {
-            return _url('ltiOverview', 'DeliveryRunner', null, ['delivery' => $delivery->getUri()]);
-        } else {
+
+        if (!$assignmentService->isDeliveryExecutionAllowed($delivery->getUri(), $user)) {
             throw new LtiException(
                 __('User is not authorized to run this delivery'),
                 LtiErrorMessage::ERROR_LAUNCH_FORBIDDEN
             );
         }
+
+        if ($user->getLaunchData()->hasVariable(self::FORCE_REDIRECT_TO_RETURN_URL)) {
+            $executionService = $this->getServiceLocator()->get(LtiDeliveryExecutionService::SERVICE_ID);
+            $executions = $executionService->getLinkedDeliveryExecutions($delivery, $currentSession->getLtiLinkResource(), $user->getIdentifier());
+            $lastDE = end($executions);
+            $url = _url(
```

Comment:
I think it would be better to use the `LtiNavigationService::getReturnUrl(...)` to build the return url